client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java

```diff
@@ -632,6 +640,108 @@ public void testProvidedNamedXContents() {
     }
 
+    public void testApiNamingConventions() throws Exception {
+        String[] notSupportedApi = new String[]{
+            "cat.aliases", "cat.allocation", "cat.count", "cat.fielddata", "cat.health", "cat.help", "cat.indices", "cat.master",
+            "cat.nodeattrs", "cat.nodes", "cat.pending_tasks", "cat.plugins", "cat.recovery", "cat.repositories", "cat.segments",
+            "cat.shards", "cat.snapshots", "cat.tasks", "cat.templates", "cat.thread_pool",
+            "cluster.allocation_explain", "cluster.pending_tasks", "cluster.remote_info", "cluster.reroute", "cluster.state",
+            "cluster.stats",
+            "count", "create", "delete_by_query", "exists_source", "get_source",
+            "indices.delete_alias", "indices.delete_template", "indices.exists_template", "indices.exists_type", "indices.get",
+            "indices.get_upgrade", "indices.put_alias", "indices.recovery", "indices.segments", "indices.shard_stores", "indices.stats",
+            "indices.upgrade", "ingest.processor_grok",
+            "mtermvectors", "nodes.hot_threads", "nodes.info", "nodes.stats", "nodes.usage", "put_script", "reindex", "reindex_rethrottle",
+            "render_search_template", "scripts_painless_execute", "search_shards",
+            "snapshot.restore", "snapshot.status",
+            "tasks.get", "termvectors", "update_by_query"
+        };
+
+        ClientYamlSuiteRestSpec restSpec = ClientYamlSuiteRestSpec.load("/rest-api-spec/api");
+        Set<String> apiSpec = restSpec.getApis().stream().map(ClientYamlSuiteRestApi::getName).collect(Collectors.toSet());
+
+        Set<String> topLevelMethodsExclusions = new HashSet<>();
+        topLevelMethodsExclusions.add("getLowLevelClient");
+        topLevelMethodsExclusions.add("close");
+
+        Map<String, Method> methods = Arrays.stream(RestHighLevelClient.class.getMethods())
+                .filter(method -> method.getDeclaringClass().equals(RestHighLevelClient.class)
+                        && topLevelMethodsExclusions.contains(method.getName()) == false)
+                .map(method -> Tuple.tuple(toSnakeCase(method.getName()), method))
+                .flatMap(tuple -> tuple.v2().getReturnType().getName().endsWith("Client")
+                        ? getSubClientMethods(tuple.v1(), tuple.v2().getReturnType()) : Stream.of(tuple))
+                .collect(Collectors.toMap(Tuple::v1, Tuple::v2));
+
+        Set<String> apiNotFound = new HashSet<>();
+
+        for (Map.Entry<String, Method> entry : methods.entrySet()) {
+            Method method = entry.getValue();
+            String apiName = entry.getKey();
+
+            assertTrue("method [" + apiName + "] is not final",
+                    Modifier.isFinal(method.getClass().getModifiers()) || Modifier.isFinal(method.getModifiers()));
+            assertTrue(Modifier.isPublic(method.getModifiers()));
+
+            if (apiName.endsWith("_async")) {
```

Comment:
Where does `_async` come from. Like with the `_` and all.

Reply:
I see! So you `snake_case`ify the method name to line up with the spec and you make sure the async version pairs up.

Reply:
Maybe worth a comment.

Reply:
yep, I will add a comment